### PR TITLE
Changes to API 

### PR DESCRIPTION
**Pull request by Aevyz**
See Devo Chat

### CODE REVIEW

**Comment by Aevyz** [comment]
#86

File: config.properties [hunk 1]
```diff
@@ -68,7 +68,8 @@ IBM_TTS_PASS: ""
 CONTEXT_GUI_ENABLED: false
 ACTION_CLIENT_SCRIPT: "/home/roboy/workspace/Roboy/src/roboy_dialog/resources/scripts/action_client.py"
 
-TELEGRAM_API_TOKENS_FILE: "resources/telegram.json"
+TELEGRAM_API_TOKENS_FILE:   "resources/telegram.json"
+EXTERNAL_API_KEY_FILE:      "resources/knowledgebase/api.key"
```
Comment (missxa) [inline config.properties]: can you pls specify what kind of API is this

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 2]
```diff
@@ -65,8 +62,7 @@
     private final static String TRANSITION_FINISHED_ANSWERING = "finishedQuestionAnswering";
     private final static String TRANSITION_LOOP_TO_NEW_PERSON = "loopToNewPerson";
     private final static String TRANSITION_LOOP_TO_KNOWN_PERSON = "loopToKnownPerson";
-    private final static String TRANSITION_SWITCH_TO_GAMING = "switchToGaming";
-    private final static int MAX_NUM_OF_QUESTIONS = 5;
+    private final static int MAX_NUM_OF_QUESTIONS = 6;
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: why is `TRANSITION_SWITCH_TO_GAMING` now removed?

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 3]
```diff
@@ -89,11 +84,10 @@ public QuestionRoboyQAState(String stateIdentifier, StateParameters params) {
 
     @Override
     public Output act() {
-        if (Math.random() < 0.1) {
-            offeredGame = true;
-            return Output.say(PhraseCollection.OFFER_GAME_PHRASES.getRandomElement());
+        if (questionsAnswered > 0) {
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: ehm? why these changes?

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 4]
```diff
@@ -103,39 +97,19 @@ public Output react(Interpretation input) {
 
     private Output reactToQuestion(Interpretation input) {
 
-        if (offeredGame) {
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: ??

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 5]
```diff
@@ -103,39 +97,19 @@ public Output react(Interpretation input) {
...
-            return Output.say(PhraseCollection.JOKES.getRandomElement()).setEmotion(RoboyEmotion.positive.getRandomElement());
-        }
-
         if (input.getTokens() != null && !(input.getTokens().contains("you") || input.getTokens().contains("your"))) {
             Linguistics.ParsingOutcome parseOutcome = input.getParsingOutcome();
-
             if (parseOutcome == null) {
                 LOGGER.error("Invalid parser outcome!");
-                return Output.useFallback();
+                return Output.say("Invalid parser outcome!");
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: don't use the hardcoded strings in states, please. they're a nightmare to track down once the system is properly deployed

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 6]
```diff
@@ -268,61 +227,43 @@ private String inferPasAnswer(Map<SemanticRole, String> pas, Roboy roboy) {
...
-                answer = extractNodeNameForPredicate(Neo4jRelationship.CREATED_BY, roboy);
             }
         } else if (matchPas(pas, new Pair(SemanticRole.PREDICATE, "do|like"), new Pair(SemanticRole.AGENT, "you"))){
-//            double prob = Math.random();
-//            if (prob < .3) {
-//                answer = extractNodeNameForPredicate(Neo4jProperty.abilities, roboy);
-//            } else if(prob < .7) {
-//                answer = extractNodeNameForPredicate(Neo4jRelationship.HAS_HOBBY, roboy);
-//            } else {
+            double prob = Math.random();
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: no, this doesnt go in

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 7]
```diff
@@ -268,61 +227,43 @@ private String inferPasAnswer(Map<SemanticRole, String> pas, Roboy roboy) {
...
-        }
         // else if {OBJ: *} -> query * -> I'm sure I know a typeof(*) called *! (Where does he live? :))
         //
         // 	if * in Neo4jRelationship.FRIEND_OF
         return answer;
     }
 
     private Neo4jRelationship inferPredicateFromObjectAnswer(String objectAnswer) {
-        if (objectAnswer.contains("hobby")) {
+        if (objectAnswer.contains("hobb")) {
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: hobby

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java [hunk 8]
```diff
@@ -368,15 +309,7 @@ private String extractNodeNameForPredicate(Neo4jRelationship predicate, Roboy ro
...
-            // check if there are multiple things in it
-            RandomList<String> properties = new RandomList<>(property.split(","));
-            //pick 2 random properties to talk about
-            String propertiesToShare = properties.getRandomElement();
-            properties.remove(propertiesToShare); // so we dont pick the same one again
-            if (!properties.isEmpty()) {
-                propertiesToShare += " and " + properties.getRandomElement();
-            }
-            return String.format(infoValues.getSuccessAnswers(predicate).getRandomElement(), propertiesToShare);
+            return String.format(infoValues.getSuccessAnswers(predicate).getRandomElement(), property);
```
Comment (missxa) [inline dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: why?

**Review by missxa — CHANGES_REQUESTED** [review]
@Aevyz please return the states you change to what they're on my branch

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 7 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: IIRC it was mentioned, that this was so that both hobby and hobbies are accepted

File: config.properties (hunk 1 above)
Comment (Aevyz) [reply config.properties]: The idea was to store all sorts of keys here, like a general key storage file for all current and future services. The idea was that `APIhandler`'s `keyGetter` internal class automatically grabs your key for you from the file. 

Would `EXTERNAL_API_KEY_LIST_FILE` be better? Or would a comment be better?

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 2 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: Copy paste from Devel, took https://github.com/Roboy/roboy_dialog/blob/devel/dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java because I know that works.

I can revert this and only copy paste the bare minimum to get the APIs to work

Regarding specific reasoning for why something was done, you may need to ask @josephbirkner

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 8 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: @josephbirkner @Waguramu

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 2 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: From what I can tell, we roboy.dialog.states.ordinaryStates.QuestionAnsweringState and used that as a basis to create the current Devo class. The Devo personality was written to be a "demo", that would answer 6 questions before doing the Rick and Morty reference at the end. APIs would be demoed in these 6 questions.

Since a game would "eat up" one of the questions, these were disabled. 


Question Answering State (for Ordinary Personality) was not touched afaik.

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 3 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: See above

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 4 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: See above

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 5 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: Should be fixed

File: dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java (hunk 6 above)
Comment (Aevyz) [reply dialog/src/main/java/roboy/dialog/states/devoStates/QuestionRoboyQAState.java]: @Waguramu @josephbirkner do you recall why we added this to this class
